File: api/schemas/tasks.py
# -*- coding: utf-8 -*-
from __future__ import annotations

from datetime import date
from typing import List, Optional

from pydantic import BaseModel


class ExtractBcRequest(BaseModel):
    tables: Optional[List[str]] = None
    output_dir: str = ""
    page_size: Optional[int] = None
    mode: str = "incremental"
    parallel: int = 1
    dry_run: bool = False
    reset_watermarks: bool = False
    checkpoint_path: str = ""
    verbose: bool = False
    notify: bool = False


class UploadBcRequest(BaseModel):
    output_dir: str = "./exports"
    dry_run: bool = False
    skip_existing: bool = False
    verbose: bool = False
    notify: bool = False


class SyncBcRequest(BaseModel):
    tables: Optional[List[str]] = None
    output_dir: str = "./exports"
    mode: str = "incremental"
    parallel: int = 1
    dry_run: bool = False
    skip_existing: bool = False
    verbose: bool = False
    notify: bool = False


class ExtractFactorialRequest(BaseModel):
    start_on: date
    end_on: date
    employees: Optional[List[int]] = None
    employee_status: str = "active"
    tables: Optional[List[str]] = None
    output_dir: str = ""
    mode: str = "full"
    parallel: int = 5
    reset_all_checkpoints: bool = False
    dry_run: bool = False
    verbose: bool = False
    notify: bool = False


class UploadFactorialRequest(BaseModel):
    output_dir: str = "./exports_factorial"
    tables: Optional[List[str]] = None
    dry_run: bool = False
    skip_existing: bool = False
    verbose: bool = False
    notify: bool = False


class SyncFactorialRequest(BaseModel):
    start_on: date
    end_on: date
    employee_status: str = "active"
    tables: Optional[List[str]] = None
    output_dir: str = "./exports_factorial"
    mode: str = "incremental"
    parallel: int = 5
    dry_run: bool = False
    skip_existing: bool = False
    verbose: bool = False
    notify: bool = False


class RunPipelineRequest(BaseModel):
    pipeline: str
    wait: bool = True
    poll_seconds: int = 15
    verbose: bool = False
    notify: bool = False


class TableStatusOut(BaseModel):
    name: str
    status: str
    detail: str = ""
    phase: str = ""


class TaskOut(BaseModel):
    id: str
    action: str
    action_label: str
    triggered_by: str
    status: str
    started_at: str
    finished_at: Optional[str] = None
    duration_seconds: float
    current_step: int
    step_labels: List[str]
    table_statuses: List[TableStatusOut]


class TaskListOut(BaseModel):
    items: List[TaskOut]

File: frontend/src/api/tasks.ts
import { apiGet, apiPost, buildQuery } from "./client";
import type { TableStatus, TaskStatus } from "./types";

export interface Task {
  id: string;
  action: string;
  action_label: string;
  triggered_by: string;
  status: TaskStatus;
  started_at: string;
  finished_at: string | null;
  duration_seconds: number;
  current_step: number;
  step_labels: string[];
  table_statuses: TableStatus[];
  log_tail: string;
}

export interface TaskFilters {
  action?: string[];
  user?: string[];
  status?: string[];
  date_from?: string;
  date_to?: string;
}

export function fetchTasks(filters: TaskFilters = {}): Promise<{ items: Task[] }> {
  return apiGet<{ items: Task[] }>(`/tasks${buildQuery(filters)}`);
}

export function fetchTask(id: string): Promise<Task> {
  return apiGet<Task>(`/tasks/${id}`);
}

export function stopTask(id: string): Promise<void> {
  return apiPost<void>(`/tasks/${id}/stop`);
}

export interface ExtractBcInput {
  tables?: string[] | null;
  output_dir?: string;
  page_size?: number | null;
  mode?: "incremental" | "full";
  parallel?: number;
  dry_run?: boolean;
  reset_watermarks?: boolean;
  checkpoint_path?: string;
  verbose?: boolean;
  notify?: boolean;
}
export function extractBc(input: ExtractBcInput): Promise<Task> {
  return apiPost<Task>("/tasks/extract-bc", input);
}

export interface UploadBcInput {
  output_dir?: string;
  dry_run?: boolean;
  skip_existing?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function uploadBc(input: UploadBcInput): Promise<Task> {
  return apiPost<Task>("/tasks/upload-bc", input);
}

export interface SyncBcInput {
  tables?: string[] | null;
  output_dir?: string;
  mode?: "incremental" | "full";
  parallel?: number;
  dry_run?: boolean;
  skip_existing?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function syncBc(input: SyncBcInput): Promise<Task> {
  return apiPost<Task>("/tasks/sync-bc", input);
}

export interface ExtractFactorialInput {
  start_on: string;
  end_on: string;
  employees?: number[] | null;
  employee_status?: "active" | "inactive" | "all";
  tables?: string[] | null;
  output_dir?: string;
  mode?: "full" | "incremental";
  parallel?: number;
  reset_all_checkpoints?: boolean;
  dry_run?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function extractFactorial(input: ExtractFactorialInput): Promise<Task> {
  return apiPost<Task>("/tasks/extract-factorial", input);
}

export interface UploadFactorialInput {
  output_dir?: string;
  tables?: string[] | null;
  dry_run?: boolean;
  skip_existing?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function uploadFactorial(input: UploadFactorialInput): Promise<Task> {
  return apiPost<Task>("/tasks/upload-factorial", input);
}

export interface SyncFactorialInput {
  start_on: string;
  end_on: string;
  employee_status?: "active" | "inactive" | "all";
  tables?: string[] | null;
  output_dir?: string;
  mode?: "incremental" | "full";
  parallel?: number;
  dry_run?: boolean;
  skip_existing?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function syncFactorial(input: SyncFactorialInput): Promise<Task> {
  return apiPost<Task>("/tasks/sync-factorial", input);
}

export interface ExtractHubspotInput {
  tables?: string[] | null;
  output_dir?: string;
  parallel?: number;
  dry_run?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function extractHubspot(input: ExtractHubspotInput): Promise<Task> {
  return apiPost<Task>("/tasks/extract-hubspot", input);
}

export interface UploadHubspotInput {
  output_dir?: string;
  tables?: string[] | null;
  dry_run?: boolean;
  skip_existing?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function uploadHubspot(input: UploadHubspotInput): Promise<Task> {
  return apiPost<Task>("/tasks/upload-hubspot", input);
}

export interface SyncHubspotInput {
  tables?: string[] | null;
  output_dir?: string;
  parallel?: number;
  dry_run?: boolean;
  skip_existing?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function syncHubspot(input: SyncHubspotInput): Promise<Task> {
  return apiPost<Task>("/tasks/sync-hubspot", input);
}

export interface RunPipelineInput {
  pipeline: string;
  wait?: boolean;
  poll_seconds?: number;
  verbose?: boolean;
  notify?: boolean;
}
export function runPipeline(input: RunPipelineInput): Promise<Task> {
  return apiPost<Task>("/tasks/run-pipeline", input);
}

export type SyncApplyDirection = "to_target" | "to_source" | "both";

export interface SyncApplyInput {
  mapping: string;
  direction: SyncApplyDirection;
  // Matching-key values (e.g. emails) to restrict this run to -- omit/undefined
  // means every pending action.
  keys?: string[];
  confirm_large_batch?: boolean;
  verbose?: boolean;
  notify?: boolean;
}
export function syncApply(input: SyncApplyInput): Promise<Task> {
  return apiPost<Task>("/tasks/sync-apply", input);
}
